File: smallpict-php/src/Crypto.php
<?php

declare(strict_types=1);

namespace SmallPict;

final class Crypto
{
    public const EMPTY_SHA256 = 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855';

    public static function sha256Hex(string $data): string
    {
        if ($data === '') {
            return self::EMPTY_SHA256;
        }
        return hash('sha256', $data);
    }

    public static function hmacSha256Hex(string $secretKey, string $stringToSign): string
    {
        return hash_hmac('sha256', $stringToSign, $secretKey);
    }

    public static function buildStringToSign(
        string $method,
        string $path,
        string $timestamp,
        string $bodyHash
    ): string {
        $cleanPath = strpos($path, '/') === 0 ? $path : '/' . $path;
        return strtoupper($method) . "\n" . $cleanPath . "\n" . $timestamp . "\n" . $bodyHash;
    }

    private function __construct()
    {
    }
}

File: smallpict-java/src/main/java/com/smallpict/Crypto.java
package com.smallpict;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;

public final class Crypto {
    public static final String EMPTY_SHA256 = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";

    private Crypto() {}

    public static String sha256Hex(byte[] data) {
        if (data == null || data.length == 0) {
            return EMPTY_SHA256;
        }
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(data);
            return bytesToHex(hash);
        } catch (Exception e) {
            throw new RuntimeException("SHA-256 algorithm unavailable", e);
        }
    }

    public static String hmacSha256Hex(String secretKey, String stringToSign) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKeySpec = new SecretKeySpec(secretKey.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
            mac.init(secretKeySpec);
            byte[] hmacBytes = mac.doFinal(stringToSign.getBytes(StandardCharsets.UTF_8));
            return bytesToHex(hmacBytes);
        } catch (Exception e) {
            throw new RuntimeException("Failed to calculate HMAC-SHA256", e);
        }
    }

    public static String buildStringToSign(String method, String path, String timestamp, String bodyHash) {
        String cleanPath = path.startsWith("/") ? path : "/" + path;
        return method.toUpperCase() + "\n" + cleanPath + "\n" + timestamp + "\n" + bodyHash;
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
}
